### ISSUE TITLE
Add integration tests for DAG

### ISSUE THREAD

**Issue by kneasle** [kneasle/sapling#36]
Currently there's no automated testing of whether or not the commands do what they are intended.

I'm not sure exactly how straightforward this is going to be; if you want to have a crack at it but are stuck with what to do, comment here and I'll give pointers!

**Comment by stokhos** [comment]
I'd like to take this one.

**Comment by kneasle** [comment]
Coool. Looks like we're running out of good-first-issues :laughing: gotta find some more...

**Comment by kneasle** [comment]
Bruh why won't github let me assign you... :angry: oh well it's yours anyway.

**Comment by kneasle** [comment]
What I think we need for these tests is a load of `(input tree, cursor location, action, output tree)` examples, where the `output tree` is the result of running the `action` on the `input tree` with the cursor at the `cursor location`.

So to make this more streamlined, we could feed all of the tests through a function like:

```rust
fn run_test(
    input_tree: TestJSON,
    cursor: CursorPath, 
    action: Action,
    exp_result: Result<TestJSON, EditErr>, 
) {
   ... 
}
```

`ast::test_json::TestJSON` is a copy of the `ast::json::JSON` enum that doesn't need to be stored in an arena. The `main` function in `main.rs` should show how to turn some `TestJSON` into a `DAG` containing the `JSON` equivalent of 5ha5 tree. Then we'd perform the `Action` and then we have two cases:
- The action was a success: we check that `exp_result` was `Ok(tree)` and that `tree` is equal to the new tree of the `DAG` (I think we might have to implement `PartialEq<JSON>` for `TestJSON`...). 
- The action returned an `EditErr` `e` and we check that `exp_result == Err(e)`

Ideally, we'd have tests to cover every possible way errors could be generated plus a few successful cases too. The gold standard is to have every line of code executed over the course of the unit tests so that bugs have nowhere to hide - but there isn't an accurate way to automatically test for coverage so just look for where errors are created and try to make test cases where that happens.

If you feel like making more cool tests then a 'fuzzer' that feeds a `DAG` random actions and just makes sure it doesn't crash.

**Comment by stokhos** [comment]
Can you give me an exam over this? I'm still learning the internals, an example would make this a lot easier for me.

**Comment by kneasle** [comment]
In the process of making an example that compiles, I realise that a surprising amount of refactoring/trait implementations have to be done to make these integration tests work.  Most of these are oversights anyway, so I'll merge these as a PR ASAP so you have a decently clean slate to work with.

Here's an example (testing the insertion of `true` into `[true, <[]>]` to get `[true, <[true]>]` (where the cursor is inside `<...>`) - this is in the example:

```rust
#[cfg(test)]
mod integration {
    use super::{cursor_path::CursorPath, EditSuccess, DAG};
    use crate::arena::Arena;
    use crate::ast::test_json::TestJSON as J;
    use crate::editor::normal_mode::Action;

    #[test]
    fn example() {
        // Inputs to the test
        let start_tree = J::Array(vec![J::True, J::Array(vec![])]);
        let start_cursor_location = CursorPath::from_vec(vec![1]);
        let action = Action::InsertChild('t');
        let expected_tree = J::Array(vec![J::True, J::Array(vec![J::True])]);
        let expected_cursor_location = CursorPath::from_vec(vec![1]);

        /* Run the test */

        // Create a `DAG` that contains the starting tree and a given cursor location (we have to
        // first create an arena for `DAG` to store its nodes).  This is very similar to Sapling's
        // startup code in `main.rs`.
        let arena = Arena::new();
        let root = start_tree.add_to_arena(&arena);
        let mut editable_tree = DAG::new(&arena, root, start_cursor_location);

        // Perform the action - we expect `editable_tree.execute_action(action)` to return
        // `(false, Ok(()))` because the action should be a success that doesn't quit Sapling.
        assert_eq!(
            editable_tree.execute_action(action),
            (
                false,
                Ok(EditSuccess::InsertChild {
                    c: 't',
                    name: "true".to_string()
                })
            )
        );

        // Check that the tree is in the expected state
        assert_eq!(expected_tree, editable_tree.root());
        assert_eq!(expected_cursor_location, editable_tree.current_cursor_path);
    }
}
```

Obviously, every test will follow this pattern (you have some initial state, an action and an expected state where 'state' is `(TestJSON, CursorPath)`) so it would make sense to push everything after `/* Run the test */` into a function - say, `run_test_success` - which takes all the `// Inputs to the test` as parameters.  Then we can write lots of unit tests very quickly by having them call `run_test_success`.

A similar function would be useful for cases where the action is expected to fail, something along the lines of
```rust
fn run_test_error(
    start_tree: TestJSON,
    start_cursor: CursorPath,
    expected_err: EditErr,
) {
    ....
}
```

Where we the same body as `run_test_success`, but the first `assert_eq!` becomes
```rust
assert_eq!(editable_tree.execute_action(action), (false, Err(expected_err)));
```
and instead of checking against an expected final state, we assert that no change to the tree has been made (i.e. the final state is the same as the inital state).

This has got pretty long, but it's mostly code and, as they say, code tells a thousand words - so I hope this helps.

**Comment by stokhos** [comment]
Ok, thank you for your help. This really helps.

**Comment by stokhos** [comment]
```rust
        let start_tree = J::Array(vec![J::True, J::Array(vec![])]);
        let start_cursor_location = CursorPath::from_vec(vec![1]);
```
There several parts that I don't fully understand. 
In `J::Array(vec![J::True, J::Array(vec![])])`, Is the first `J::Array` root node, and it has 2 children, `J::True`, and `J::Array(vec![])`? 
`CursorPath::from_vec(vec![1])` If we replaced the  `1` to `0`, this means the cursor would be at the first element `J::True`?

```rust
assert_eq!(expected_tree, editable_tree.root());
```
In here  `editable_tree.root()` will return the root node, so if I call `editable_tree.cursor()`, this will return the node where my cursor is at, and its children if there are any. Is this correct?

**Comment by kneasle** [comment]
> `CursorPath::from_vec(vec![1])` If we replaced the `1` to `0`, this means the cursor would be at the first element `J::True`? Which number should I fill in to indicate the cursor is at root node?

The way `CursorPath`s are used to traverse the tree uses something like the following Python code:
```python
# Start out pointing at the root
node = root()
# Read off each index from `cursor_path`
for index in cursor_path:
    # Use each index to step down to a child of `node`
    node = node.children[index]
# Once we've run out of indices, this must be the node that the CursorPath was pointing at
return node
```

So the root node has an empty path `[]` - there's even a special constructor for constructing it: `CursorPath::root()`.  I thought there was some half-decent documentation for `CursorPath` but apparently there isn't... I'll make an issue and get on to that soon.

---

> In here `editable_tree.root()` will return the root node, so if I call `editable_tree.cursor()`, this will return the node where my cursor is at, and its children if there are any. Is this correct?

Yes, that is correct - `DAG::root()` returns a reference to the current root node, and `DAG::cursor()` returns a reference to the node that is under the cursor.  Both of these will point to memory owned by the `Arena` that the specific `DAG` is using to store nodes, and both of these will contain references to their descendant trees (so debug printing `DAG::root()` will end up debug printing the entire tree, despite the fact that `DAG::root()` is a pointer to only the root node).

**Comment by stokhos** [comment]
@kneasle 
I'm not sure how to test `EditErr::InsertError`, `EditErr::DeleteError` and `EditErr::CharNotANode`.

I think first two might be related to lacking methods to inserting a node directly, like `DAG.insert(node: Node)`.

The third one might be redundant with `EditErr::CannotBeChild`.

**Comment by stokhos** [comment]
No `char` stands for `JSON::Field` enum for now.

From signature I dont understand why it has to follow : The first `Ref` must be a [`Str`](JSON::Str), and the second is any.
```rust
    /// A JSON object field.  The first `Ref` must be a [`Str`](JSON::Str), and the second is any
    /// JSON object
    Field([&'arena JSON<'arena>; 2])
```

**Comment by kneasle** [comment]
> I'm not sure how to test `EditErr::InsertError`, `EditErr::DeleteError` and `EditErr::CharNotANode`.

I think that in the current set-up, none of these _can_ be created - but I don't think they're redundant; there might be some obscure edge case where they suddenly occur and be useful.

----

> No `char` stands for `JSON::Field` enum for now.

It's true that `JSON::Field` has no `char` right now, but that's largely because currently fields are implicitly created when inserting nodes.  So it doesn't make sense for fields to have a `char` (it also brings up questions like: what should default value of a `JSON::Field` be?).

> From signature I dont understand why it has to follow : The first `Ref` must be a `Str`, and the second is any.
> 
> ```rust
>     /// A JSON object field.  The first `Ref` must be a [`Str`](JSON::Str), and the second is any
>     /// JSON object
>     Field([&'arena JSON<'arena>; 2])
> ```

I used to have the signature of this something like
```rust
Field { key: &'arena JSON<'arena>, value: &'arena JSON<'arena> }
```
and an invariant that the LHS should be a `JSON::Str`.  But the rest of the code relies on being able to represent the children as a slice and that wasn't possible in that implementation without `unsafe` code.  So I moved to the slice-like representation and what I'll admit is a bit of a cryptic comment.

**Comment by stokhos** [comment]
Do I need to wait until you merged #53 ?

**Comment by kneasle** [comment]
> Do I need to wait until you merged #53 ?

No I don't think so - #53 didn't change the interface of DAG at all. If anything I think _this_ should be merged first, to make sure that my refactoring hasn't accidentally introduced any bugs or changed the behaviour of DAG. 

That will also give me the opportunity to add a 'regression test' for the 'delete last child; undo; redo' crash that I fixed on stream.
